WIP: add bolt backend in gluten

WangGuangxin opened the pull request.

Commit:
add bolt backend in gluten

kevinwilfong committed:
[VL] Support mapping columns by position index for ORC and Parquet files (#10697)

afterincomparableyum commented:
Thanks for your hardwork, really appreciate it @WangGuangxin . I know Velox does certain queries where the result does not match with what vanilla spark outputs. Is this the same with Bolt? I think it would be nice to include some docs about any potential mismatches for users to be cautious about

taiyang-li committed:
test commit

taiyang-li committed:
Add parent project information to pom.xml

taiyang-li committed:
fix style

kexianda committed:
Refine build script

* add make argument for specifing  Bolt's version(branch/tag)
* minor fix

kexianda committed:
Add Dockerfile for bolt backend

guhaiyan0221 committed:
register SparkExprToSubfieldFilterParser

guhaiyan0221 committed:
fix style

kexianda committed:
Refine Makefile

* remove hardcoded java path
* install jdk17 for centos dockerfile

guhaiyan0221 committed:
Optimize out-of-the-box parameters

WangGuangxin committed:
overwrite batchsize default value

FelixYBW commented:
build with s3 failed:


`make release ENABLE_S3=True`

```

CMake Error at CMakeLists.txt:362 (find_package):
  By not providing "FindAWSSDK.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "AWSSDK", but
  CMake did not find one.

  Could not find a package configuration file provided by "AWSSDK" with any
  of the following names:

    AWSSDKConfig.cmake
    awssdk-config.cmake

  Add the installation prefix of "AWSSDK" to CMAKE_PREFIX_PATH or set
  "AWSSDK_DIR" to a directory containing one of the above files.  If "AWSSDK"
  provides a separate development package or SDK, be sure it has been
  installed.

```

taiyang-li committed:
add make arrow instruction

taiyang-li committed:
add docker instructions

guhaiyan0221 committed:
fix S3 compile error

guhaiyan0221 committed:
add Bolt.md

guhaiyan0221 committed:
add bolt-spark-configuration.md

guhaiyan0221 committed:
add BoltStageResourceAdj.md

guhaiyan0221 committed:
bolt-backend-generator-function-support.md

guhaiyan0221 committed:
add aggregate-function/scalar-function/window-function/write-configuration.md

guhaiyan0221 committed:
add bolt-function-development-guide.md BoltFileSystem.md BoltLocalCache.md

guhaiyan0221 committed:
add velox-to-bolt-migration-guide.md

guhaiyan0221 committed:
add bolt-quick-start.md

kexianda committed:
align gtest version wiht bolt

WangGuangxin committed:
Remove outputType init logic

FelixYBW commented:
Velox PRs submited by Gluten committers and related:
https://github.com/bytedance/bolt/discussions/191

kexianda committed:
fix: symbols conflicts with other JNI libraries

Currently to make sure llvm IR can call c/c++ code, we added a customized library
loader which loads bolt backend with RTLD_GLOBAL flag.
Since bolt_backend is a shared libraries, it will expose all the dynamic
symbol table to Java process, and will cause symbols conflict if the user has other native libraries.

Use version script to control the symbols exposure.

How to check:
```
readelf -d -s --dyn-syms libbolt.so | grep jit_*
```

Commit:
cp commit: Fix heap oom caused by too large InSet

(cherry picked from commit b0cec2c56938484371c8f20916cb7e7c09c6e19c)

ZacBlanco committed:
Add support for paimon from master

This change introduces all of the necessary paimon-specific
changes that were included in olap/gluten's master branch
but were not ported over when switching to OSS gluten's
main branch

The following changes were made

- PaimonScanTransformer is now AbstractPaimonScanTransformer
in order to support a other backend's implementation for paimon.
Bolt-specific features are now included in an extended class
BoltPaimonScanTransformer to facilitate the bolt-specific
requirements.

- Added support for adding extension info from the scan
transformer into the advanced_extension field for a ReadRel.
Also added protos for the paimon-specific advanced extension
which are automatically generated and compiled for both Java
and C++ versions.
- The above also coincides with adding support for passing
the "tableParameters" field of a HiveTableHandle when
converting the substrait plan into bolt.

- Added protos specific to paimon splits in order to serialize
and deserialize paimon-specific split information from gluten
into bolt. Previously this was done in a hacky way using
strings and comma-separated lists. This new version uses
protobuf both on the Java and C++ side to communicate
paimon-specific split information via a LocalFiles' file_format
field oneof definition in algebra.proto. One of the new options
added is for "PaimonReadOptions" which contains the required
paimon information.

- Added a bolt-specific paimon suite which mirrors all of
the previous test cases from the master branch.

fzhedu committed:
fix: avoid flatten twice in write

Signed-off-by: fangzhuhe <fangzhuhe@bytedance.com>

markjin1990 committed:
[fix] pass the full tables schema when creating HiveTableHandle for orc files to make sure Bolt can correctly handle legacy orc files

guhaiyan0221 committed:
fix compilation error caused by renaming RegisterGCSFileSystem.h to RegisterGcsFileSystem.h

kexianda committed:
Remove __cxa_throw hook in gluten

VvanFalleaves committed:
Remove unused data type cases in BoltBackend

Removed handling for StructType, ArrayType, and MapType in data type matching.

zhangxffff committed:
remove unused sort_before_repartition for round robin shuffle

ZacBlanco committed:
fix: dont push down paimon metadata column filters

fzhedu committed:
add spark.gluten.sql.columnar.backend.bolt.shuffle.check.ratio

Signed-off-by: fangzhuhe <fangzhuhe@bytedance.com>

zhangxffff committed:
fix: guard against null task_ in WholeStageResultIterator::collectMetrics and hashagg config

zjuwangg committed:
[Bolt] Support Spark map_from_entries function

zjuwangg committed:
[Bolt] Enable map_from_arrays function with Bolt backend

zjuwangg committed:
feat: support sequence fucntion

zjuwangg committed:
[Bolt]Add trunc support

zjuwangg committed:
[Bolt] Support Spark format_number function

Register FormatNumber expression mapping for Bolt backend so that
Spark's format_number function can be offloaded to Bolt. Added a
unit test in ScalarFunctionsValidateSuite and updated the supported
scalar function list.

zjuwangg committed:
fix ut for map_from_entries

zhangxffff committed:
fix: first/last intermediate struct uses row_constructor

For first/last the intermediate is (value, valueSet). The bolt fix
guarantees the struct is always non-null and valueSet is always non-null
(false or true), so row_constructor is used instead of
row_constructor_with_null. This prevents a single empty partition's
partial from overwriting a real value with NULL via the updateNull path.

zjuwangg committed:
fix ut for sequence/format_number  ut

zjuwangg committed:
Fix FIRST_WIN in spark config

FelixYBW commented:
@WangGuangxin have you rebased after @taiyang-li's refactor?

guhaiyan0221 commented:
> @WangGuangxin have you rebased after @taiyang-li's refactor?

we created a new PR after refactor. https://github.com/apache/gluten/pull/12454
